### PR TITLE
[SC-337] March 06 Spell

### PR DESCRIPTION
**Pull request by barrutko**

### CODE REVIEW

File: src/proposals/20240306/SparkEthereum_20240306.t.sol [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+        /*******************/
+
+        ReserveConfig memory wethConfigAfter = wethConfigBefore;
+
+        wethConfigAfter.interestRateStrategy = _findReserveConfigBySymbol(allConfigsAfter, 'WETH').interestRateStrategy;
+
+        wethConfigAfter.ltv                  = 82_00;
+        wethConfigAfter.liquidationThreshold = 83_00;
+
+        _validateReserveConfig(wethConfigAfter, allConfigsAfter);
```
Comment (hexonaut) [inline src/proposals/20240306/SparkEthereum_20240306.t.sol:225]: Missing IRM update check for WETH. Add below: `_validateInterestRateStrategy`

File: src/SparkTestBase.sol [hunk 2]
```diff
@@ -334,6 +343,34 @@ abstract contract SparkEthereumTestBase is SparkTestBase {
         _assertPaused(WETH, true);
     }
 
+    function _assertBorrowCapConfig(address asset, uint256 max, uint256 gap, uint256 increaseCooldown) internal {
+        (uint256 _max, uint256 _gap, uint256 _increaseCooldown,,) = capAutomator.borrowCapConfigs(asset);
```
Comment (hexonaut) [inline src/SparkTestBase.sol:347]: Solidity doesn't pack these, but the sizing is actually `uint48`. They should be updated to this just to be extra cautious.

File: src/SparkTestBase.sol [hunk 3]
```diff
@@ -334,6 +343,34 @@ abstract contract SparkEthereumTestBase is SparkTestBase {
...
+        assertEq(_max,              max);
+        assertEq(_gap,              gap);
+        assertEq(_increaseCooldown, increaseCooldown);
+    }
+
+    function _assertBorrowCapConfigNotSet(address asset) internal {
+        (uint256 _max, uint256 _gap, uint256 _increaseCooldown,,) = capAutomator.borrowCapConfigs(asset);
```
Comment (hexonaut) [inline src/SparkTestBase.sol:354]: Same here. `uint48`.

File: src/SparkTestBase.sol [hunk 4]
```diff
@@ -334,6 +343,34 @@ abstract contract SparkEthereumTestBase is SparkTestBase {
...
+        assertEq(_max,              0);
+        assertEq(_gap,              0);
+        assertEq(_increaseCooldown, 0);
+    }
+
+    function _assertSupplyCapConfig(address asset, uint256 max, uint256 gap, uint256 increaseCooldown) internal {
+        (uint256 _max, uint256 _gap, uint256 _increaseCooldown,,) = capAutomator.supplyCapConfigs(asset);
```
Comment (hexonaut) [inline src/SparkTestBase.sol:361]: Same here. `uint48`.

File: src/SparkTestBase.sol [hunk 5]
```diff
@@ -334,6 +343,34 @@ abstract contract SparkEthereumTestBase is SparkTestBase {
...
+        assertEq(_max,              max);
+        assertEq(_gap,              gap);
+        assertEq(_increaseCooldown, increaseCooldown);
+    }
+
+    function _assertSupplyCapConfigNotSet(address asset) internal {
+        (uint256 _max, uint256 _gap, uint256 _increaseCooldown,,) = capAutomator.supplyCapConfigs(asset);
```
Comment (hexonaut) [inline src/SparkTestBase.sol:368]: Same here. `uint48`.

File: src/proposals/20240306/SparkEthereum_20240306.t.sol [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+    function testCapAutomatorCapUpdates() public {
+        GovHelpers.executePayload(vm, payload, executor);
+
+        _assertAutomatedCapsUpdate(RETH);
+        _assertAutomatedCapsUpdate(SDAI);
+        _assertAutomatedCapsUpdate(USDC);
+        _assertAutomatedCapsUpdate(USDT);
+        _assertAutomatedCapsUpdate(WBTC);
+        _assertAutomatedCapsUpdate(WETH);
+        _assertAutomatedCapsUpdate(WSTETH);
```
Comment (hexonaut) [inline src/proposals/20240306/SparkEthereum_20240306.t.sol:397]: I would also just run over each asset with hardcoded values to make sure the results make sense.

Ex) `assertEq(weth.configuration.getSupplyCap(), 388_000)` (or whatever it is actually 238k existing + 150k gap)

File: src/proposals/20240306/SparkEthereum_20240306.t.sol [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+
+        _validateReserveConfig(rethConfigAfter, allConfigsAfter);
+
+        IDefaultInterestRateStrategy rethIRSAfter = IDefaultInterestRateStrategy(
+            rethConfigAfter.interestRateStrategy
+        );
+
+        _validateInterestRateStrategy(
+            address(rethIRSAfter),
+            address(rethIRSAfter),
```
Comment (lucas-manuel) [inline src/proposals/20240306/SparkEthereum_20240306.t.sol:178]: Not a big deal but these can be `rethConfigAfter.interestRateStrategy` or `rethIRSAfter` can be declared as an address

File: src/proposals/20240306/SparkEthereum_20240306.t.sol [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+            asset:            WSTETH,
+            max:              3_000,
+            gap:              100,
+            increaseCooldown: 12 hours
+        });
+    }
+
+    function testCapAutomatorCapUpdates() public {
+        GovHelpers.executePayload(vm, payload, executor);
+
```
Comment (lucas-manuel) [inline src/proposals/20240306/SparkEthereum_20240306.t.sol:390]: Can we change this to iterate through all assets in the reserves list? Should still work

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: src/proposals/20240306/SparkEthereum_20240306.t.sol (hunk 8 above)
Comment (hexonaut) [reply src/proposals/20240306/SparkEthereum_20240306.t.sol:402]: done

**Review by lucas-manuel — APPROVED** [review]

**Comment by lucas-manuel** [comment]
Good to deploy

**Review by hexonaut — APPROVED** [review]

**Review by lucas-manuel — APPROVED** [review]

File: src/proposals/20240306/SparkEthereum_20240306.t.sol [hunk 9]
```diff
@@ -400,13 +400,27 @@ contract SparkEthereum_20240306Test is SparkEthereumTestBase {
...
+        assertEq(pool.getReserveData(USDT).configuration.getSupplyCap(),   30_000_000);
+        assertEq(pool.getReserveData(USDT).configuration.getBorrowCap(),   3_098_915);
+        assertEq(pool.getReserveData(WBTC).configuration.getSupplyCap(),   5_000);
+        assertEq(pool.getReserveData(WBTC).configuration.getBorrowCap(),   182);
+        assertEq(pool.getReserveData(WETH).configuration.getSupplyCap(),   389_229);
+        assertEq(pool.getReserveData(WETH).configuration.getBorrowCap(),   126_809);
+        assertEq(pool.getReserveData(WSTETH).configuration.getSupplyCap(), 735_128);
+        assertEq(pool.getReserveData(WSTETH).configuration.getBorrowCap(), 207);
+        assertEq(pool.getReserveData(DAI).configuration.getSupplyCap(),    0);
+        assertEq(pool.getReserveData(DAI).configuration.getBorrowCap(),    0);
```
Comment (barrutko) [inline src/proposals/20240306/SparkEthereum_20240306.t.sol]: In this PR, assets were always ordered in alphabetical order. Here DAI breaks out.
Not a huge deal though.

**Review by barrutko — COMMENTED** [review]
Reviewed the deploy and the changes made on top of my commits.
One minor stylistic nitpick, but other than that LGTM.
(I can't officially approve my own PR, so just commenting)

File: src/proposals/20240306/SparkEthereum_20240306.t.sol (hunk 9 above)
Comment (hexonaut) [reply src/proposals/20240306/SparkEthereum_20240306.t.sol]: Let's just keep as is unless you feel strongly. Feel free to merge.

Comment (barrutko) [reply src/proposals/20240306/SparkEthereum_20240306.t.sol]: As is is fine

**Pull request merged**